minitask2/benh/robot_control.py

```diff
@@ -0,0 +1,289 @@
+#!/usr/bin/env python
+import rospy
+import tf
+import random
+
+from nav_msgs.msg import Path, Odometry
+from geometry_msgs.msg import Twist, PoseStamped, Pose
+from sensor_msgs.msg import LaserScan
+
+from abc import abstractmethod
+from math import pi as PI, radians, degrees, sqrt, cos
+
+THREE_METERS = 3
+
+SAMPLE_WINDOW = 20  # Degrees
+
+TURNING_ERROR_THRESHOLD = radians(10)
+TURN_GAIN = 0.5
+TURN_MAX_SPEED = 1
+TURN_MIN_SPEED = 0.01
+
+FORWARD_GAIN = 0.4
+FORWARD_MAX_SPEED = 0.3
+FORWARD_MIN_SPEED = 0.1
+
+
+class RobotBehaviour(object):
+    '''
+    Abstract class that other robot behaviours inherit from.
+    '''
+
+    def __init__(self, vel_pub):
+        self.vel_pub = vel_pub
+        self.move_twist = Twist()
+        self.name = "!!!"
+
+    def sample_scan(self, scandata, target, window_size=SAMPLE_WINDOW):
+        '''
+        Returns the minimum value in the given window_size, centered on target
+        '''
+        assert type(window_size) == int, "window_size should be an integer"
+        offset = window_size / 2
+        # Was going to use min() and list slicing but its a pain with wrapping around
+        min_val = scandata.ranges[target - offset]
+        for i in range(target - offset + 1, target + offset):
+            min_val = min(scandata.ranges[i], min_val)
+        return min_val
```

Comment:
Would it be possible to use `min(scandata.ranges[target-offset+1, target+offset])` instead of `for` loop?

Reply:
Yeah that was what I originally tried. If you have something like `min(scandata.ranges[target - offset + 1:target + offset])` and the target is directly ahead (`0`), offset is maybe `5` then it'll evaluate to `min(scandata.ranges[-4:5])` which initially seems like the right thing.

However the -4 gets converted to the index `len(scandata.ranges) - 4` which means your final line is something like `min(scandata.ranges[356:5])`, now because the start index > end index it will return `[]` (empty list). I could do something where if the first index is less than 0 get the subarray from that to the end of the list then the start of the list to the original end index but I reckon thats pretty messy.